src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+     */
+    export interface RangeToExtract {
+        readonly targetRange?: TargetRange;
+        readonly errors?: Diagnostic[];
+    }
+
+    /*
+     * Scopes that can store newly extracted method
+     */
+    export type Scope = FunctionLikeDeclaration | SourceFile | ModuleBlock | ClassLikeDeclaration;
+
+
+    /**
+     * Result of 'extractRange' operation for a specific scope.
+     * Stores either a list of changes that should be applied to extract a range or a list of errors
+     */
+    export interface ExtractResultForScope {
+        readonly scope: Scope;
+        readonly scopeDescription: string;
+        readonly changes?: FileTextChanges[];
+        readonly errors?: Diagnostic[];
+    }
+
+    /**
+     * getRangeToExtract takes a span inside a text file and returns either an expression or an array
+     * of statements representing the minimum set of nodes needed to extract the entire span. This
+     * process may fail, in which case a set of errors is returned instead (these are currently
+     * not shown to the user, but can be used by us diagnostically)
+     */
+    export function getRangeToExtract(sourceFile: SourceFile, span: TextSpan): RangeToExtract {
+        const length = span.length || 0;
+        // Walk up starting from the the start position until we find a non-SourceFile node that subsumes the selected span.
+        // This may fail (e.g. you select two statements in the root of a source file)
+        let start = getParentNodeInSpan(getTokenAtPosition(sourceFile, span.start, /*includeJsDocComment*/ false), sourceFile, span);
+        // Do the same for the ending position
+        let end = getParentNodeInSpan(findTokenOnLeftOfPosition(sourceFile, textSpanEnd(span)), sourceFile, span);
+
+        // We'll modify these flags as we walk the tree to collect data
+        // about what things need to be done as part of the extraction.
+        let facts = RangeFacts.None;
+
+        if (!start || !end) {
+            // cannot find either start or end node
+            return { errors: [createFileDiagnostic(sourceFile, span.start, length, Messages.CannotExtractFunction)] };
+        }
+
+        if (start.parent !== end.parent) {
+            // handle cases like 1 + [2 + 3] + 4
+            // user selection is marked with [].
+            // in this case 2 + 3 does not belong to the same tree node
+            // instead the shape of the tree looks like this:
+            //          +
+            //         / \
+            //        +   4
+            //       / \
+            //      +   3
+            //     / \
+            //    1   2
+            // in this case there is no such one node that covers ends of selection and is located inside the selection
+            // to handle this we check if both start and end of the selection belong to some binary operation
+            // and start node is parented by the parent of the end node (because binary operators are left associative)

```

Comment:
I don't think it matters in this case because (AFAICT) you can't get into the situation described when the selected operator is an assignment expression, IOW `1 + 2 = 3 + 4` is a parse error so it's not really possible to get into the state where the associativity matters. Would be interested in a test case that might test this hypothesis more fully